content/articles/automation_lessons.md

```diff
@@ -0,0 +1,110 @@
+title: Automation Lessons
+tags: software engineering, collaboration
+author: cdent
+created: 2025-02-26
+
+```intro
+I initially wrote the following for work colleagues, but sharing more widely
+in case it might apply elsewhere.
+
+A lesson from long ago that has stuck with me and been proven valuable over and
+over:
+
+> When creating software, ensure there are working harnesses for testing,
+> installation, and upgrade first.
+```
+
+Over the years as development technologies have matured the lesson has evolved
+to assert the harnesses must be _automated_ and that the automation _must run
+before changes are merged_.
+
+We've all probably learned this main lesson and the various side lessons many
+times over, but frequent review is required because...
+
+...Throughout my experience with our previous teams we've frequently ignored
+this lesson for a variety of reasons, all of which are at least inconvenient
+but most of the time invalid:
+
+1. We're in too much of a hurry to add tests, we need to get this out now.
+2. This is too complicated to automate.
+3. This is too time consuming to run in pull request workflows.
+4. This is too resource consuming to run in pull request workflows.
+
+Most of these reasons only feel correct when viewed from the perspective of the
+individual developer who needs to do the work to make things go, or is waiting
+on CI for the pull request to merge.
+
+For other developers, other teammates, or the customer (who might be a "real"
+customer, or someone internal) these reasons don't have much relevance.
+
+If you don't test now, then it is the other people who end up doing the testing
+for you, finding issues. Turnaround time on resolving those issues is vastly
+higher than it would be otherwise, the other person has to do extra work to get
+what they want and you have to do extra work to understand the problem. Instead
+you could experience the problem yourself, sooner, and not cascade the wasted
+time to others. That wasted time adds up, exponentially, and any time savings
+you thought you were getting by pushing things out quickly are shrunk,
+sometimes dramatically.
+
+If you don't automate the tests or the installation, you then have to spend
+time doing those things by hand. You do. Your teammates do. The customer does.
+And no matter how much you think "oh we're only going to do this by hand once
+or twice" that always turns out to be a lie.
+
+Even when it doesn't turn out to be a lie, eventually you'll need test and
+installation automation. Paying the price up front has many advantages:
+
+-   There is a significant risk that there won't be time later.
+
+-   Existing test and installation automation provides an exceptionally valuable
+    design constraint when adding or modifying features. If something is hard to
```

Comment:
```suggestion
    [design constraint](page://articles/constraints) when adding or modifying features. If something is hard to
```

I'm not sure the slant of [that article](https://prepitaph.org/articles/constraints/) makes for a suitable reference here, but its title - one of your own active lessons, of course - seems relevant.

Reply:
yup, the title applies, but the connection is a bit buried at the end, so not sure